### PR TITLE
Add saturating_add() and saturating_sub() to integers

### PR DESCRIPTION
**Pull request by Shnatsel**
NEON has all these ops natively. x86 has u8/i8/u16/i16, but 32-bit and 64-bit element widths need to be emulated; that emulation is the source of most of the complexity.

### CODE REVIEW

**Comment by Shnatsel** [comment]
LLVM 23 pessimizes `srai(a, 31) ^ i32::MAX` and turns it into two constant broadcasts and a blend. Workarounds are available but I'm not going to special-case i32 further, the code is complicated enough.

File: fearless_simd_tests/tests/harness/ops/saturating_add.rs
Comment (LaurenzV) [inline fearless_simd_tests/tests/harness/ops/saturating_add.rs:1]: Hmm I know we are a bit inconsistent with this right now, but could we change the formatting here so it's not spread out over so many lines? For example more like we do it here:

Image

Comment (LaurenzV) [reply fearless_simd_tests/tests/harness/ops/saturating_add.rs:1]: (Same for the other test file.)

File: fearless_simd_gen/src/ops.rs [hunk 1]
```diff
@@ -971,12 +978,28 @@ const INT_OPS: &[Op] = &[
         OpSig::Binary,
         "Add two vectors element-wise, wrapping on overflow.",
     ),
+    Op::new(
+        "saturating_add",
+        OpKind::VecTraitMethod,
+        OpSig::Binary,
+        "Add two vectors element-wise, returning the maximum value on overflow.\n\n\
```
Comment (LaurenzV) [inline fearless_simd_gen/src/ops.rs]: Or minimum for negative numbers, I presume?

**Review by LaurenzV — APPROVED** [review]

File: fearless_simd_gen/src/ops.rs (hunk 1 above)
Comment (Shnatsel) [reply fearless_simd_gen/src/ops.rs]: Yes. I've struggled to clearly and succinctly convey this. The standard library only says "saturating at the numeric bounds instead of overflowing." which is not very clear either.

File: fearless_simd_tests/tests/harness/ops/saturating_add.rs
Comment (Shnatsel) [reply fearless_simd_tests/tests/harness/ops/saturating_add.rs:1]: I tried putting i8::MAX into `max` and `i8::MIN` into `min` variables before opening this PR, but that didn't affect the formatting at all.

If you know of a way to achieve this, please enlighten me.

Comment (LaurenzV) [reply fearless_simd_tests/tests/harness/ops/saturating_add.rs:1]: Oh hmm, that's weird... if it doesn't work then nevermind, but I wonder why it works in the other cases.

Comment (Shnatsel) [reply fearless_simd_tests/tests/harness/ops/saturating_add.rs:1]: Is this any better? seems like a toss-up to me.

```rust
// Binding boundary values to constants keeps rustfmt from putting every array
// element on its own line.
const I8_MIN: i8 = i8::MIN;
const I8_MAX: i8 = i8::MAX;
const I8_MIN_P1: i8 = I8_MIN + 1;
const I8_MIN_P10: i8 = I8_MIN + 10;
const I8_MAX_M1: i8 = I8_MAX - 1;
const I8_MAX_M10: i8 = I8_MAX - 10;

const U8_MAX: u8 = u8::MAX;
const U8_MAX_M1: u8 = U8_MAX - 1;
const U8_MAX_M5: u8 = U8_MAX - 5;
const U8_MAX_M10: u8 = U8_MAX - 10;
const U8_HALF: u8 = U8_MAX / 2;
const U8_HALF_P1: u8 = U8_HALF + 1;

const I16_MIN: i16 = i16::MIN;
const I16_MAX: i16 = i16::MAX;
const I16_MIN_P1: i16 = I16_MIN + 1;
const I16_MINP10: i16 = I16_MIN + 10;
const I16_MAX_M1: i16 = I16_MAX - 1;
const I16_MAXM10: i16 = I16_MAX - 10;

const U16_MAX: u16 = u16::MAX;
const U16_MAX_M1: u16 = U16_MAX - 1;
const U16_MAX_M5: u16 = U16_MAX - 5;
const U16_MAXM10: u16 = U16_MAX - 10;
const U16_HALF: u16 = U16_MAX / 2;
const U16_HALFP1: u16 = U16_HALF + 1;

const I32_MIN: i32 = i32::MIN;
const I32_MAX: i32 = i32::MAX;
const I32_MIN_P1: i32 = I32_MIN + 1;
const I32_MINP10: i32 = I32_MIN + 10;
const I32_MINP20: i32 = I32_MIN + 20;
const I32_MAX_M1: i32 = I32_MAX - 1;
const I32_MAXM10: i32 = I32_MAX - 10;

const U32_MAX: u32 = u32::MAX;
const U32_MAX_M1: u32 = U32_MAX - 1;
const U32_MAX_M5: u32 = U32_MAX - 5;
const U32_MAXM10: u32 = U32_MAX - 10;
const U32_HALF: u32 = U32_MAX / 2;
const U32_HALFP1: u32 = U32_HALF + 1;

const I64_MIN: i64 = i64::MIN;
const I64_MAX: i64 = i64::MAX;
const I64_MIN_P1: i64 = I64_MIN + 1;
const I64_MAX_M1: i64 = I64_MAX - 1;

const U64_MAX: u64 = u64::MAX;
const U64_MAX_M1: u64 = U64_MAX - 1;
const U64_MAX_M5: u64 = U64_MAX - 5;
const U64_MAXM10: u64 = U64_MAX - 10;
const U64_HALF: u64 = U64_MAX / 2;
const U64_HALFP1: u64 = U64_HALF + 1;

#[simd_test]
fn saturating_add_i8x16<S: Simd>(simd: S) {
    let a = i8x16::from_slice(
        simd,
        &[
            I8_MAX, I8_MIN, I8_MAX_M1, I8_MIN_P1, 40, -40, 50, -50, 0, 0, I8_MAX, I8_MIN,
            I8_MAX_M10, I8_MIN_P10, 1, -1,
        ],
    );
    let b = i8x16::from_slice(
        simd,
        &[
            1, -1, 1, -1, 20, -20, -75, 75, I8_MAX, I8_MIN, 0, 0, 9, -9, -1, 1,
        ],
    );
    assert_eq!(
        *a.saturating_add(b),
        [
            I8_MAX, I8_MIN, I8_MAX, I8_MIN, 60, -60, -25, 25, I8_MAX, I8_MIN, I8_MAX, I8_MIN,
            I8_MAX_M1, I8_MIN_P1, 0, 0,
        ]
    );
}
```

and then all the following lines are compact

Comment (Shnatsel) [reply fearless_simd_tests/tests/harness/ops/saturating_add.rs:1]: specifically I can't use `MAX - 1` or indentifiers longer than 10 characters without blowing up the line count

Comment (LaurenzV) [reply fearless_simd_tests/tests/harness/ops/saturating_add.rs:1]: Meh... It's fine, you can just leave it as is 😄
